[css-anchor-1] anchor() fails to account for grid-area

fantasai opened the pull request:
#### a5da259342fb2ea31834a8131b5310f5c595978e
<pre>
[css-anchor-1] anchor() fails to account for grid-area
<a href="https://bugs.webkit.org/show_bug.cgi?id=295924">https://bugs.webkit.org/show_bug.cgi?id=295924</a>
<a href="https://rdar.apple.com/155823420">rdar://155823420</a>

Reviewed by Alan Baradlay.

PositionedLayoutConstraints relies on grid area information cached during layout,
but we are not in the middle of layout while computing anchor() insets.

This patch revises PositionedLayoutConstraints to call for a direct computation
of the grid area range, and removes all the extra methods that interact with the
now unnecessary out-of-flow grid-area cache.

It also reorders some of the area computation internals to avoid accessing data
that isn&apos;t valid prior to layout, see <a href="https://commits.webkit.org/298027@main">https://commits.webkit.org/298027@main</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-position-grid-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-try-grid-001-expected.txt:

Pass more tests.

* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
(WebCore::PositionedLayoutConstraints::captureGridArea):

Call new API for directly computing grid-area LayoutRange.

* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutBlock):
(WebCore::RenderGrid::layoutOutOfFlowBox):
(WebCore::RenderGrid::columnAxisOffsetForGridItem const):
(WebCore::RenderGrid::rowAxisOffsetForGridItem const):
(WebCore::RenderGrid::gridAreaRangeForOutOfFlow const):
(WebCore::RenderGrid::gridAreaBreadthForOutOfFlowGridItem): Deleted.
(WebCore::RenderGrid::logicalOffsetForOutOfFlowGridItem const): Deleted.
(WebCore::RenderGrid::gridAreaRowRangeForOutOfFlow const): Deleted.
(WebCore::RenderGrid::gridAreaColumnRangeForOutOfFlow const): Deleted.
(WebCore::RenderGrid::gridAreaPositionForOutOfFlowGridItem const): Deleted.
(WebCore::RenderGrid::gridAreaPositionForGridItem const): Deleted.
* Source/WebCore/rendering/RenderGrid.h:

Adjust out-of-flow grid-area size computation method to return a LayoutRange,
rename it to gridAreaRangeForOutOfFlow(), and delete all now-unused methods
and data types.

Originally-landed-as: 297834@main (fb8ef69f7e7a). <a href="https://rdar.apple.com/155823420">rdar://155823420</a>
Canonical link: <a href="https://commits.webkit.org/298644@main">https://commits.webkit.org/298644@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4ca09be520c14039ffd2eba1081702080cd0539d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114249 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33996 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24458 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120415 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64983 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/45f045dc-da46-461f-9c86-6fee8d280510) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116138 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34625 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42557 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86824 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41769 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e4a20ca3-2eb1-4a8a-b35c-f6cb5aab8e72) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117197 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27574 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102611 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67214 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26755 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20737 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64107 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96948 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20854 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123630 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41267 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30775 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95657 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41643 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98813 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95440 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24621 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40578 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18410 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37346 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41147 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46656 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40749 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44055 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42500 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
```diff
@@ -173,18 +173,11 @@ void PositionedLayoutConstraints::captureGridArea()
-        m_containingRange = *range;
-        m_containingInlineSize = range->size();
+        m_containingRange = gridContainer->gridAreaRangeForOutOfFlow(m_renderer, Style::GridTrackSizingDirection::Columns);
+        m_containingInlineSize = m_containingRange.size();
```
Seems like we always set `m_containingInlineSize` to the grid area's column size. Maybe we want to factor that out of the if else?

fantasai replied:
No because it's faster to take the cached value from `m_containingRange` than to recalculate it.

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -2307,97 +2302,43 @@ Grid& RenderGrid::currentGrid()
         end = positions[endLine];
         // These vectors store line positions including gaps, but we shouldn't consider them for the edges of the grid.
         std::optional<LayoutUnit> availableSizeForGutters = availableSpaceForGutters(direction);
+        int lastLine = numTracks(direction);
```
Should this be unsigned?

fantasai replied:
No, because then the comparison below will complain about the comparison with signed line numbers.

(Note I didn't change this line, I just moved it.)

fantasai committed:
[css-anchor-1] anchor() fails to account for grid-area
https://bugs.webkit.org/show_bug.cgi?id=295924
rdar://155823420

Reviewed by Alan Baradlay.

PositionedLayoutConstraints relies on grid area information cached during layout,
but we are not in the middle of layout while computing anchor() insets.

This patch revises PositionedLayoutConstraints to call for a direct computation
of the grid area range, and removes all the extra methods that interact with the
now unnecessary out-of-flow grid-area cache.

It also reorders some of the area computation internals to avoid accessing data
that isn't valid prior to layout, see https://commits.webkit.org/298027@main

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-position-grid-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-try-grid-001-expected.txt:

Pass more tests.

* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
(WebCore::PositionedLayoutConstraints::captureGridArea):

Call new API for directly computing grid-area LayoutRange.

* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutBlock):
(WebCore::RenderGrid::layoutOutOfFlowBox):
(WebCore::RenderGrid::columnAxisOffsetForGridItem const):
(WebCore::RenderGrid::rowAxisOffsetForGridItem const):
(WebCore::RenderGrid::gridAreaRangeForOutOfFlow const):
(WebCore::RenderGrid::gridAreaBreadthForOutOfFlowGridItem): Deleted.
(WebCore::RenderGrid::logicalOffsetForOutOfFlowGridItem const): Deleted.
(WebCore::RenderGrid::gridAreaRowRangeForOutOfFlow const): Deleted.
(WebCore::RenderGrid::gridAreaColumnRangeForOutOfFlow const): Deleted.
(WebCore::RenderGrid::gridAreaPositionForOutOfFlowGridItem const): Deleted.
(WebCore::RenderGrid::gridAreaPositionForGridItem const): Deleted.
* Source/WebCore/rendering/RenderGrid.h:

Adjust out-of-flow grid-area size computation method to return a LayoutRange,
rename it to gridAreaRangeForOutOfFlow(), and delete all now-unused methods
and data types.

Originally-landed-as: 297834@main (fb8ef69f7e7a). rdar://155823420
Canonical link: https://commits.webkit.org/298644@main
